core/domain/exp_domain.py

```diff
@@ -474,6 +474,273 @@ def __init__(self, change_list):
+        """
+        if (version_start < self._version_start or
+                version_end > self._version_end):
+            raise ValueError('Requested version range is out-of-bounds')
+        slice_start, slice_end = (
+            version_start - self._version_start,
+            version_end - self._version_start)
+        return self._versioned_states[slice_start:slice_end]
+
+    def get_multi_names(self, version_start, version_end):
+        """Returns the state names corresponding to the given half-open range of
+        exploration versions.
+
+        Args:
+            version_start: int. The lower bound of versions to get (inclusive).
+            version_end: int. The upper bound of versions to get (exclusive).
+
+        Returns:
+            list(str). The state names corresponding to the given half-open
+            range of exploration versions.
+
+        Raises:
+            ValueError. The input range is out-of-bounds.
+        """
+        return [
+            state_name for state_name, _ in self.get_multi(
+                version_start, version_end)
+        ]
+
+    def get_multi_contents(self, version_start, version_end):
+        """Returns the state contents corresponding to the given half-open range
+        of exploration versions.
+
+        Args:
+            version_start: int. The lower bound of versions to get (inclusive).
+            version_end: int. The upper bound of versions to get (exclusive).
+
+        Returns:
+            list(state_domain.State). The state contents corresponding to the
+            given half-open range of exploration versions.
+
+        Raises:
+            ValueError. The input range is out-of-bounds.
+        """
+        return [
+            state_content for _, state_content in self.get_multi(
+                version_start, version_end)
+        ]
+
+    @staticmethod
+    def _default_state_equality_predicate(unused_state1, unused_state2):
+        """Always returns True."""
+        return True
+
+
+class ExplorationStatesHistory(python_utils.OBJECT):
+    """Organizes the history of changes made to each state of an exploration."""
+
+    def __init__(self, exps, exp_version_diffs, state_equality_predicate=None):
+        """Constructs the exploration's state history.
+
+        Args:
+            exps: list(Exploration). The exploration's representations, ordered
+                by version. Items should begin at version 1 and contain no gaps.
+            exp_version_diffs: list(ExplorationVersionsDiff). The specific
+                changes made to the exploration at each corresponding version.
+            state_equality_predicate:
+                callable(state_domain.State, state_domain.State) -> bool | None.
+                Returns True when two states are "equal". The predicate is used
+                to enforce the following invariant: all versions of the state in
+                a mapping are equivalent. If None, then uses a predicate that
```

Comment:
Ditto, why do you need a default arg at all? Have the caller always specify it so that the developer is forced to think about it. 

Perhaps give examples to illustrate possible scenarios so that they understand what to pass in and what its effect is ("For example, if state_equality_predicate is lambda x, y: ..., then this means ...".) (And ditto above.)